### PR TITLE
refactor(authority): message update chain info to update single chains

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
- The `update_chain_info` message has been refactored to update/add a single chain ,instead of providing the entire list as `ChainInfo` 
  - The user is now required to provide a json file with the details of the chain to be updated/added.
    - If the chain already exists, the details will be updated.
    - If the chain does not exist, it will be added to the list of chains and saved to the store.
 - A new transaction type `RemoveChainInfo` has also been added to remove a chain from the list of chains.
    - It accepts the chain-id of the chain to be removed as a parameter.

Closes https://github.com/zeta-chain/node/issues/2815

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Summary by CodeRabbit

- **New Features**
	- Introduced a command to remove chain information via the `zetacored` CLI.
	- Added a new message type, `MsgRemoveChainInfo`, for removing chain data.
	- Enhanced the `MsgUpdateChainInfo` functionality to streamline updates for individual chains.

- **Bug Fixes**
	- Improved error handling for unauthorized removal attempts and missing chain information.

- **Documentation**
	- Updated CLI documentation to include new command options.
	- Enhanced API documentation with definitions for new message types and responses.

- **Tests**
	- Added comprehensive tests for the new `MsgRemoveChainInfo` functionality and updated existing tests for clarity.

### CODE REVIEW

File: proto/zetachain/zetacore/authority/tx.proto [hunk 1]
```diff
@@ -53,8 +55,17 @@ message MsgUpdatePoliciesResponse {}
 // MsgUpdateChainInfo defines the MsgUpdateChainInfo service.
 message MsgUpdateChainInfo {
   string creator = 1;
-  ChainInfo chain_info = 2 [ (gogoproto.nullable) = false ];
+  pkg.chains.Chain chain = 2 [ (gogoproto.nullable) = false ];
```
Comment (gartnera) [inline proto/zetachain/zetacore/authority/tx.proto:58]: Will this change break backwards compatibility?

Comment (kingpinXD) [reply proto/zetachain/zetacore/authority/tx.proto:58]: Yes, after this release, you need to submit transactions with `Chain` and not `ChainInfo.` 
I am not sure what you mean by compatibility; the chain info object stored on the state would still be the same

Comment (gartnera) [reply proto/zetachain/zetacore/authority/tx.proto:58]: Basically will a similar issue to this happen: https://github.com/zeta-chain/node/issues/2835

Comment (kingpinXD) [reply proto/zetachain/zetacore/authority/tx.proto:58]: No, this changes a field in the transaction.
I added a comment to track this though , as I am still not very sure what is the root cause of original issue

Comment (swift1337) [inline proto/zetachain/zetacore/authority/tx.proto:58]: This change breaks backward compatibility

File: x/authority/keeper/msg_server_remove_chain_info.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	updatedChainInfo := RemoveChain(chainInfo, msg.ChainId)
+	k.SetChainInfo(ctx, updatedChainInfo)
+	return &types.MsgRemoveChainInfoResponse{}, nil
+}
+
+// RemoveChain removes the chain with the given chainID from the chainInfo
+// If the chain does not exist, the chainInfo is returned as is
+func RemoveChain(chainInfo types.ChainInfo, chainID int64) types.ChainInfo {
```
Comment (lumtis) [inline x/authority/keeper/msg_server_remove_chain_info.go:37]:
```suggestion
func removeChain(chainInfo types.ChainInfo, chainID int64) types.ChainInfo {
```
Not sure we specifically need a sub function here as the logic is pretty small

File: proto/zetachain/zetacore/authority/tx.proto (hunk 1 above)
Comment (lumtis) [inline proto/zetachain/zetacore/authority/tx.proto:58]: Let's use reversed for 2 and use a new field even if it's a message

File: x/authority/keeper/msg_server_update_chain_info.go [hunk 3]
```diff
@@ -24,8 +25,27 @@ func (k msgServer) UpdateChainInfo(
 	if err != nil {
 		return nil, errors.Wrap(types.ErrUnauthorized, err.Error())
 	}
-	// set chain info
-	k.SetChainInfo(ctx, msg.ChainInfo)
 
+	chainInfo := types.ChainInfo{}
+	updatedChainInfo := false
+
+	cf, found := k.GetChainInfo(ctx)
```
Comment (lumtis) [inline x/authority/keeper/msg_server_update_chain_info.go:32]:
```suggestion
	foundChainInfo, found := k.GetChainInfo(ctx)
```
Let's use explicit variable names

File: x/authority/keeper/msg_server_update_chain_info.go [hunk 4]
```diff
@@ -24,8 +25,27 @@ func (k msgServer) UpdateChainInfo(
 	if err != nil {
 		return nil, errors.Wrap(types.ErrUnauthorized, err.Error())
 	}
-	// set chain info
-	k.SetChainInfo(ctx, msg.ChainInfo)
 
+	chainInfo := types.ChainInfo{}
+	updatedChainInfo := false
```
Comment (lumtis) [inline x/authority/keeper/msg_server_update_chain_info.go:30]:
```suggestion
	chainInfoExist := false
```
More descriptive imo as it is updated in any case

File: x/authority/keeper/msg_server_remove_chain_info.go (hunk 2 above)
Comment (kingpinXD) [reply x/authority/keeper/msg_server_remove_chain_info.go:37]: This is a bit easier to test. I'd like to keep it

File: proto/zetachain/zetacore/authority/tx.proto (hunk 1 above)
Comment (swift1337) [reply proto/zetachain/zetacore/authority/tx.proto]: I agree with Alex. It should not be possible to change **existing** proto fields. It should be `reserved 2` + `pkg.chains.Chain chain = 3`

Comment (kingpinXD) [reply proto/zetachain/zetacore/authority/tx.proto]: Both changes have already been done 
https://github.com/zeta-chain/zeta-node/blob/89db4b85296a2b32fc13d17b07d0f51650298de4/proto/zetachain/zetacore/authority/tx.proto#L57-L58

**Review by swift1337 — APPROVED** [review]

**Review by lumtis — APPROVED** [review]

**Pull request merged**